### PR TITLE
feat: move @bazel/worker

### PR DESCRIPTION
**Pull request by thesayyn**
still not finalized.

TODO
- docs
- generating .d.ts output of the API.

### CODE REVIEW

File: js/private/worker/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        abortionMap.set(request.request_id, abortController)
+
+        const response = new blaze.worker.WorkResponse({
+            request_id: request.request_id,
+        })
+
+        const outputBuffer = Buffer.alloc(0)
+        const outputStream = new Writable({ write: outputBuffer.write })
+
+        implementation({
```
Comment (thesayyn) [inline js/private/worker/src/index.ts:49]: no use of await here to achieve a non-blocking loop.  the loop has to be blocked on stdin only.

File: js/private/worker/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                    responseBytes.byteLength
+                )
+                process.stdout.write(
+                    Buffer.concat([responseSizeBytes, responseBytes])
+                )
+            })
+    }
+}
+
+export function shouldRunAsWorker(args: string[]): boolean {
```
Comment (thesayyn) [inline js/private/worker/src/index.ts:80]: i feel like there is a better name for this function.

File: js/private/worker/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        const request = blaze.worker.WorkRequest.deserialize(current!)
+
+        if (request.cancel) {
+            abortionMap.get(request.request_id)?.abort()
+            continue
+        }
+
+        const abortController = new AbortController()
+        abortionMap.set(request.request_id, abortController)
```
Comment (thesayyn) [inline js/private/worker/src/index.ts:40]: not to be confused with child abortion...

File: js/private/worker/src/index.ts (hunk 1 above)
Comment (thesayyn) [reply js/private/worker/src/index.ts:49]: necessary for multiplexing and cancellation

File: js/private/worker/src/index.ts (hunk 2 above)
Comment (jbedard) [reply js/private/worker/src/index.ts:89]: When would the "bazel worker" not be run as a worker?

**Comment by jbedard** [comment]
Is this editing it or only moving it? Can we see a diff of any changes you made?

File: js/private/worker/src/index.ts (hunk 2 above)
Comment (thesayyn) [reply js/private/worker/src/index.ts:89]: when you put `--spawn_strategy=TsProject=local|sandboxed` into your .bazelrc

**Comment by thesayyn** [comment]
> Is this editing it or only moving it? Can we see a diff of any changes you made?

we got a completely different API now.

**Review by gregmagolan — APPROVED** [review]

**Comment by gregmagolan** [comment]
> > Is this editing it or only moving it? Can we see a diff of any changes you made?
> 
> we got a completely different API now.

Looks like `size.ts` hasn't changed and `worker_protoco.ts` is generated so the new impl is in `index.ts`

File: js/private/worker/src/index.ts (hunk 2 above)
Comment (alexeagle) [reply js/private/worker/src/index.ts:89]: may as well name it after the flag you read so you don't have to invent a name.

`isPersistentWorker` ?

**Comment by gregmagolan** [comment]
Is the gc.ts --expose_gc special handling no longer needed: https://github.com/bazelbuild/rules_nodejs/blob/stable/packages/worker/src/gc.ts

I vaguely recall that being added at some point

File: js/private/worker/src/index.ts (hunk 2 above)
Comment (thesayyn) [reply js/private/worker/src/index.ts:89]: that's better. Thanks

File: js/private/worker/src/index.ts (hunk 1 above)
Comment (gregmagolan) [reply js/private/worker/src/index.ts:55]: 👍

**Pull request merged**